src/data/data.cc

```diff
@@ -146,6 +145,21 @@ void MetaInfo::SetInfo(const char* key, const void* dptr, DataType dtype, size_t
   }
 }
 
+// Explicit template instantiation.
+template BatchSet<SparsePage> DMatrix::GetBatches<SparsePage>(PageType);
+
+template<typename T>
+BatchSet<T> DMatrix::GetBatches(PageType page_type) {
```

Comment:
this approach defeats the very purpose of abstraction - not to mention more maintenance when new types evolve. it presents a tight coupling between the types when the base type has to know about every derived type, which is what polymorphism was intended to avoid.